### PR TITLE
Fix #32: Implement exception handling.

### LINKED ISSUE #32: Support try-catch-finally
Now, we should be able to support try-catch-finally using [Wasm EH](https://github.com/WebAssembly/exception-handling/blob/master/proposals/exception-handling/Exceptions.md) and [availability of java.lang.Throwable](https://github.com/tanishiking/scala-wasm/pull/28)

We should be able to catch the JavaScript exception in future(?), but the initial focus should be on catching the exceptions thrown in Wasm.

We'll include only one exception tag within the module, whose type is `java.lang.Throwable`. Whenever an exception is thrown, it will always compile to `throw 0` with an operand of type (or subtype of) `java.lang.Throwable`.

```lisp
(tag $tag (param (ref null $java.lang.Throwable)))) ;; whose tag idx is 0
```

and the catch clause in Wasm always catch the j`ava.lang.Throwable` with `catch 0`, and then validate the exception's type. If none of catch clauses (in Scala) caught an exception, rethrow the exception.

### try-catch

```scala
// ExceptionA extends Exception
// ExceptionB extends Exception
try {
  throw Exception()
} catch (e: ExceptionA) {
} catch (e: ExceptionB) {
}
```

should be compiled to

```lisp
(local $ex (ref null $java.lang.Throwable)
try  
    ;; construct the Exception and push to the stack
    throw 0
catch 0
    local.tee $ex ;; thrown exception
    ref.test $ExceptionA ;; test if the thrown exception is a subtype of ExceptionA
    if ;; catch(e: ExceptionA) { ... }
        ;; ...
    else
        local.get $ex
        ref.test $ExceptionB
        if ;; catch (e: ExceptionB) { ... }
            ;; ...
            else ;; if none of catch clauses catch the exception
                local.get $ex
                throw 0 ;; rethrow it
            end
        end
    end)
```

### try-catch-finally

```scala
try {
    throw Exception()
} catch (e: Exception) {
} finally { println("hello") }
```

```lisp
block $finally
    try
        try
            ;; construct exception
            throw 0
        catch 0
             local.tee $ex
             ref.test $Exception
             if
                 ;; catch(e: Exception) { ... }
             else ;; catch clauses couldn't capture the exeption rethrow
                 local.get $ex
                 throw 0
              end
              ;; reach here if one of the catch clause captured the exception
              br $finally ;; jump to outside of the block and run finally block
         end
         unreachable
    catch 0 ;; if the exception isn't caught by any catch clauses, do something in finally clause and rethrow the exception.
          ;; println("hello") (finally clause)
          ;; push the caught exception to stack
          throw 0
     end
     unreachable
end ;; of block
;; println("hello") finally clause
```

note: it seems try-catch-finally is compiled to something above in IR :heart:

```scala
    val x = try {
        val x = 1
    } catch {
        case e: ExceptionA => js.Dynamic.global.console.log("finally")
        case e: ExceptionB => js.Dynamic.global.console.log("finally")
    } finally {
        js.Dynamic.global.console.log("finally")
    }

TryFinally(
  TryCatch(
    Block(
      VarDef(LocalIdent(LocalName<x>),NoOriginalName,IntType,false,IntLiteral(1)),
      Undefined()
    ),
    LocalIdent(LocalName<e>),
    NoOriginalName,
    If(
      IsInstanceOf(
        VarRef(LocalIdent(LocalName<e>)),
        ClassType(ClassName<sample.ExceptionA>)
      ),
      Block(
        VarDef(LocalIdent(LocalName<e$3>),OriginalName(e),ClassType(ClassName<sample.ExceptionA>),false,AsInstanceOf(VarRef(LocalIdent(LocalName<e>)),ClassType(ClassName<sample.ExceptionA>))),
        JSMethodApply(JSGlobalRef(console),StringLiteral(log),List(Apply(org.scalajs.ir.Trees$ApplyFlags@0,LoadModule(ClassName<scala.scalajs.js.Any$>),MethodIdent(MethodName<fromString;Ljava.lang.String;Lscala.scalajs.js.Any>),List(StringLiteral(finally)))))
      ),
      If(
        IsInstanceOf(VarRef(LocalIdent(LocalName<e>)),ClassType(ClassName<sample.ExceptionB>)),
        Block(
          VarDef(LocalIdent(LocalName<e$2>),OriginalName(e),ClassType(ClassName<sample.ExceptionB>),false,AsInstanceOf(VarRef(LocalIdent(LocalName<e>)),ClassType(ClassName<sample.ExceptionB>))),
          JSMethodApply(JSGlobalRef(console),StringLiteral(log),List(Apply(org.scalajs.ir.Trees$ApplyFlags@0,LoadModule(ClassName<scala.scalajs.js.Any$>),MethodIdent(MethodName<fromString;Ljava.lang.String;Lscala.scalajs.js.Any>),List(StringLiteral(finally)))))),
        Throw(VarRef(LocalIdent(LocalName<e>)))
      )
    )
  ),
  JSMethodApply(JSGlobalRef(console),StringLiteral(log),List(Apply(org.scalajs.ir.Trees$ApplyFlags@0,LoadModule(ClassName<scala.scalajs.js.Any$>),MethodIdent(MethodName<fromString;Ljava.lang.String;Lscala.scalajs.js.Any>),List(StringLiteral(finally)))))
)
```

### PR DESCRIPTION
**Pull request by sjrd**
Unfortunately, we cannot write unit tests for `TryCatch` and `TryFinally` yet. The latest Node.js nightly build is still using an old version of v8 that does not support the latest spec for Wasm exception handling. However, it possible to run the `sample` from a web browser when it contains `try..catch` and `try..finally`.

### CODE REVIEW

File: wasm/src/main/scala/ir2wasm/LibraryPatches.scala [hunk 1]
```diff
@@ -31,8 +31,13 @@ object LibraryPatches {
       }
     }
 
+    val leanerJSExceptionIRFile =
+      org.scalajs.linker.backend.emitter.PrivateLibHolder.files.find { irFile =>
```
Comment (tanishiking) [inline wasm/src/main/scala/ir2wasm/LibraryPatches.scala:35]: - https://github.com/scala-js/scala-js/blob/80eafa9ecb54f6763c0df32350c9c6cd8e7343c5/linker/js/src/main/scala/org/scalajs/linker/backend/emitter/PrivateLibHolder.scala#L23-L34
- https://github.com/scala-js/scala-js/blob/80eafa9ecb54f6763c0df32350c9c6cd8e7343c5/linker-private-library/src/main/scala/scala/scalajs/js/JavaScriptException.scala#L17-L21

**Review by tanishiking — APPROVED** [review]
LGTM!

Thank you for updating the EH-spec related stuffs, I didn't know that there was an update on the spec recently https://github.com/WebAssembly/exception-handling/pull/283 😅 

IIUC, this implementation is able to catch the exception thrown from JS like

> JS throws something else -> Scala.js catches it wrapped in a js.JavaScriptException
> https://www.scala-js.org/doc/interoperability/exceptions.html

am I right?

**Comment by sjrd** [comment]
> IIUC, this implementation is able to catch the exception thrown from JS like
> 
> > JS throws something else -> Scala.js catches it wrapped in a js.JavaScriptException
> > https://www.scala-js.org/doc/interoperability/exceptions.html
> 
> am I right?

Unfortunately not yet. For that we will need a not-yet-approved addition to the exception handling feature. See https://github.com/WebAssembly/exception-handling/pull/269

Currently all the wrap/unwrap thing does is allow to throw non-`Throwable` the way JavaScript would do.

File: wasm/src/main/scala/ir2wasm/WasmExpressionBuilder.scala [hunk 2]
```diff
@@ -1328,6 +1326,73 @@ private class WasmExpressionBuilder private (
...
+     * pairs that cross its boundary. In case there is a `Return` inside the
+     * `try` block targeting a `Labeled` block around this `TryFinally`, the
+     * `finally` block is by-passed.
+     */
+
+    val resultType = TypeTransformer.transformResultType(t.tpe)(ctx)
+    val resultLocals = resultType.map(fctx.addSyntheticLocal(_))
+
+    fctx.block() { doneLabel =>
+      fctx.block(Types.WasmExnRef) { catchLabel =>
```
[the remark this answers is not in the document]
Comment (sjrd) [reply wasm/src/main/scala/ir2wasm/WasmExpressionBuilder.scala:1360]: `try_table` is always a try..catch, yes. We have to implement the `finally` behavior manually on top of that.

(That's not uncommon for low-level bytecode formats. The JVM does the same.)

**Comment by tanishiking** [comment]
> Currently all the wrap/unwrap thing does is allow to throw non-Throwable the way JavaScript would do.

Gotcha, thanks!

**Pull request merged**